feat(oke): Create OKE nodes with Ubuntu images.

uhryniuk opened the pull request:
OKE LA has specific instructions for how to pull the image and use it when creating different kinds of k8s nodes.

simpoir commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+- ``kubectl`` installed (Self-Managed only).
+
+
+Download an Ubuntu image
```
Although this is not technically wrong, the whole download/upload can by bypassed by pointing directly to the image in our storage bucket. I think this is significantly nicer. e.g.

```
oci compute image import from-object-url https://intcanonical.objectstorage.us-phoenix-1.oci.customer-oci.com/p/vpqQtYASl8IooEZ_sxfKDnzUkF1b-3lQmXPC_rXf4zARQYoW7ncE8BGGxNqdUuGa/n/intcanonical/b/oke-shared/o/jammy/oke-1.29/20240820/daily-ubuntu-amd64-minimal-22.04-jammy-v20240820-oke-1.29.img ...
```

hammerstefan replied:
Good catch.  This significantly reduces the number of steps and the time it takes for a user to import.

uhryniuk replied:
Very cool, I'll look at merging the 2 sections now

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+    --wait-for-state AVAILABLE
+
+
+Creating OKE Nodes with Ubuntu Images
```
```suggestion
Create OKE nodes with Ubuntu Images
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+
+The following steps on creating nodes assumes you have an existing OKE cluster on Oracle Cloud, but it is not required to have existing nodes. If you don't have an OKE cluster prepared then Oracle's documentation for `creating a cluster`_ is a good place to start.
+
+Creating Managed OKE nodes with Ubuntu
```
```suggestion
Create managed OKE nodes with Ubuntu
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+Creating Managed OKE nodes with Ubuntu
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Managed nodes are node instances whose lifecycle is managed by the OKE service. Since this is a Limited Availability release of Ubuntu images for OKE, you can only create managed nodes through the Oracle Cloud API (``oci`` CLI or SDK). The ability to create managed from nodes from the Oracle Cloud UI will be added at a later.
```
```suggestion
Managed nodes are node instances whose lifecycle is managed by the OKE service. Since this is a Limited Availability release of Ubuntu images for OKE, you can only create managed nodes through the Oracle Cloud API (``oci`` CLI or SDK). The ability to create managed nodes from the Oracle Cloud UI will be added later.
```

hammerstefan commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+------------------------
+
+
+To download an image, select a version from the `available releases <#available-releases>`_ and make note of the image path for the image you choose. The image path conforms to the following format:
```
I think we need a section on how to interpret the JSON listing provided by #available-release links.  It is a sorted array where the most recent image will be at the end (with the last two entries being the latest image and the latest image manifest respectively).

I'd also like to add a CLI snippet, along the lines of 
```
 curl https://intcanonical.objectstorage.us-phoenix-1.oci.customer-oci.com/p/vpqQtYASl8IooEZ_sxfKDnzUkF1b-3lQmXPC_rXf4zARQYoW7ncE8BGGxNqdUuGa/n/intcanonical/b/oke-shared/o/ | jq ".[][-2] | .name"
```
to automate getting the latest.

Design note (somewhat off topic here):
We could also create a `<suite>/oke-<version>/current` prefix in the bucket, that contains the latest image released, which might be easier to navigate/inspect.

hammerstefan commented:
I think it'd be good to use `tabs` to split "Using Console" and "Using CLI" for each of the relevant sections. See https://canonical-oracle.readthedocs-hosted.com/en/latest/_sources/oracle-how-to/use-bastion-to-access-VM.rst.txt for examples

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+    --node-metadata='{"user_data": "'"$(base64 user-data.yaml)"'"}'
+
+
+View the node pool status in Oracle Cloud by navigating to :guilabel:`Kubernetes Clusters (OKE)` and click on your cluster, then select :guilabel:`Resources` > :guilabel:`Node pools` and click on the latest node pool.
```
Try to use 'select' instead of 'click' whenever possible. (Since people might not necessarily be using a mouse)

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+
+Everything will be running as expected when the :guilabel:`Kubernetes node condition` and :guilabel:`Node state` of all the nodes are labelled :guilabel:`Ready`.
+
+Creating Self-Managed OKE nodes with Ubuntu
```
```suggestion
Create self-managed OKE nodes with Ubuntu
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+Creating Self-Managed OKE nodes with Ubuntu
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following instructions assumes that you have configured your OKE cluster to work with self-managed nodes. If you have not done this, refer to the Oracle documentation for `working with self-managed nodes`_
```
```suggestion
The following instructions assume that you have configured your OKE cluster to work with self-managed nodes. If you have not done this, refer to the Oracle documentation for `working with self-managed nodes`_
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+
+.. code:: bash
+  
+  kubectl cluster-info
```
Is that the whole command? Where is the OKE cluster specified? Is that part of what follows in the output of the command? If so, it would be nice to see that too.

uhryniuk replied:
That's the command as a whole, kubectl defaults to using a config file that we assume the user has set up when we refer to 'configuring kubectl'.

It returns some command returns the cluster status in stdout, would it be worth adding that to the code block?

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+.. code:: bash
+
+   kubectl config view --raw -o json | jq -r '.clusters[0].cluster.certificate-authority-data'
+
```
Maybe include a statement about how to determine your cluster index in cases where multiple clusters are configured.

uhryniuk replied:
Added `--minify` \[0\] flag so the index isn't needed anymore.

In summary, `kubectl` can have multiple "contexts" which represent different clusters the `kubectl` client can point to/communicate with. Only _one_ context may be active at a time, hence why we needed the index before to select the active option. With `--minify` it trims the data to only return the data for the currently active context.
 
Refs:
- \[0\]: https://kubernetes.io/docs/reference/kubectl/generated/kubectl_config/kubectl_config_view/

hammerstefan replied:
I think that's good and less confusing. Thank you!

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+
+   kubectl config view --raw -o json | jq -r '.clusters[0].cluster.certificate-authority-data'
+
+Then obtain the ``Kubernetes API private endpoint`` from Oracle Cloud by navigating to :guilabel:`Kubernetes Cluster (OKE)` then select your cluster. Be sure to copy only the IP, not the port.
```
```suggestion
Then obtain the ``Kubernetes API private endpoint`` from Oracle Cloud by navigating to :guilabel:`Kubernetes Cluster (OKE)` and selecting your cluster. Be sure to copy only the IP, not the port.
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+
+   oci ce cluster get --cluster-id <cluster-id> | jq -r '.data.endpoints.private-endpoint' | cut -d ":" -f1
+
+Using your previously obtained values, insert them in the following example and save it with the name ``user-data.yaml``.
```
```suggestion
Use these obtained values (certificate-data and private-endpoint) in the following example and save it as ``user-data.yaml``.
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+     content: <certificate-data>
+
+
+Now, create the instance behind the self-managed node in Oracle Cloud by navigating to :guilabel:`Compute` > :guilabel:`Instance` and click :guilabel:`Create Instance`. Next, click :guilabel:`Change Image` and select :guilabel:`My Images` then click the Ubuntu image recently registered. 
```
```suggestion
Now, create the self-managed node in Oracle Cloud by navigating to :guilabel:`Compute` > :guilabel:`Instance` and click :guilabel:`Create Instance`. Next, click :guilabel:`Change Image` and select :guilabel:`My Images` then click the Ubuntu image recently registered. 
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+Further references
+------------------
+
+For more information about oci CLI and managing self-managed nodes on your cluster, refer to the Oracle Documentation:
```
Use sentence case for the link titles

uhryniuk commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+
+Ubuntu images are available for worker nodes on Oracle Kubernetes Engine (OKE) in Oracle Cloud. Currently there are only a select number of suites and Kubernetes versions supported due to this being a Limited Availability release. 
+
+Available Releases
```
I understand the confusion, we're intending to have more rows here once we add support for focal and jammy along with more k8s versions too. Having said that, definitely going to change to 'List of images' still.

Should sentence case be used with the top level title, like ``Deploy OKE Nodes using Ubuntu Images`` -> ``Deploy OKE nodes using Ubuntu images``?

a-dubs commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+
+   oci ce cluster get --cluster-id <cluster-id> | jq -r '.data.endpoints.private-endpoint' | cut -d ":" -f1
+
+Use these obtained values (certificate-data and private-endpoint) in the following example and save it as ``user-data.yaml``.
```
```suggestion
Use these obtained values (certificate-data and private-endpoint) in the following example cloud-config and save it as ``user-data.yaml``.
```

a-dubs commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+
+Now, create the self-managed node in Oracle Cloud by navigating to :guilabel:`Compute` > :guilabel:`Instance` and click :guilabel:`Create Instance`. Next, click :guilabel:`Change Image` and select :guilabel:`My Images` then click the Ubuntu image recently registered. 
+
+Setup the cloud-init for the instance by clicking :guilabel:`Show advanced options` then select :guilabel:`Paste cloud-init script`, where you should be able to paste your completed cloud-init script (the one saved in ``user-data.yaml``).
```
```suggestion
Setup cloud-init for the instance by selecting :guilabel:`Show advanced options`, then selecting :guilabel:`Paste cloud-init script`, and then paste your completed cloud-init script there (the one saved in ``user-data.yaml``).
```

a-dubs commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+     content: <certificate-data>
+
+
+Now, create the self-managed node in Oracle Cloud by navigating to :guilabel:`Compute` > :guilabel:`Instance` and click :guilabel:`Create Instance`. Next, click :guilabel:`Change Image` and select :guilabel:`My Images` then click the Ubuntu image recently registered. 
```
```suggestion
Now, create the self-managed node in Oracle Cloud by navigating to :guilabel:`Compute` > :guilabel:`Instance` and then selecting :guilabel:`Create Instance`. Next, select :guilabel:`Change Image` and select :guilabel:`My Images`, and then select the Ubuntu image you recently registered. 
```

a-dubs commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+  
+  kubectl cluster-info
+
+Next, the self-managed node will need a custom cloud-init script which needs some specific values, namely a Kubernetes certificate from the OKE cluster and the Kubernetes API private endpoint.
```
```suggestion
Next, the self-managed node will need a custom cloud-init user-data script which needs some specific values, namely a Kubernetes certificate from the OKE cluster and the Kubernetes API private endpoint.
```

a-dubs commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,250 @@
+
+The following instructions assume that you have configured your OKE cluster to work with self-managed nodes. If you have not done this, refer to the Oracle documentation for `working with self-managed nodes`_
+
+Before adding a self-managed node, ensure you have configured ``kubectl`` to your OKE cluster with the following command. This process will be easier if ``kubectl`` is configured for a single OKE cluster.
```
should this be:
```suggestion
Before adding a self-managed node, ensure you have configured ``kubectl`` on your OKE cluster with the following command. This process will be easier if ``kubectl`` is configured for a single OKE cluster.
```

uhryniuk replied:
Personally, I think it makes more sense with 'for'. ``kubectl`` is a client that talks to your cluster(s), the OKE service itself doesn't have any configuration to use it.

uhryniuk commented:
Tried to address everything:
- combine upload & register
- Links added for register image docs
- links added for oci cli usage (imports)
- Moved away from click, using select now.
- Fixed heading to use sentence case
- Launch mode context added
- Explain JSON in PAR, add additional 'latest' command (revisit)
- network plugin availability context added.
- tabs added for upload/register, managed and self-managed
- Loads of small grammar fixes.

uhryniuk commented on .custom_wordlist.txt:
```diff
@@ -193,6 +193,8 @@ oci
 OCI
 OCID
 OCIR
+oke
```
removed, but added kubernetes and uri, as linter complained about me using it in a `.. code:: bash` block.

lmk if I formatted something wrong, i can remove these 2 new entries.

hammerstefan reviewed (approved):
Great work!

a-dubs reviewed (approved):
LGTM

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,273 @@
+- ``kubectl`` installed (Self-Managed only).
+
+
+Finding an Ubuntu image
```
```suggestion
Find an Ubuntu image
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,273 @@
+Finding an Ubuntu image
+-----------------------
+
+Select a version from the `available releases <#available-releases>`_. The images are listed as JSON in ascending order, therefore the latest image will be at the bottom. After viewing the `available releases <#available-releases>`_, make note of the image path for the image you choose. The image path conforms to the following format:
```
```suggestion
Select a version from the `available releases <#available-releases>`_. The images are listed as JSON in ascending order, therefore the latest image will be at the bottom. Make note of the image path for the image you choose. The image path conforms to the following format:
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,273 @@
+
+    .. group-tab:: Using console
+    
+        Start the registration process Oracle Cloud by navigating to :guilabel:`Compute` > :guilabel:`Custom Images` and select :guilabel:`Import Image`. Select :guilabel:`Import from an Object Storage URL`, then paste the `available releases <#available-releases>`_ location link with your concatenated image path into the :guilabel:`Object Storage URL` field. The URL format pasted should conform  ...
```
```suggestion
        Start the registration process in Oracle Cloud by navigating to :guilabel:`Compute` > :guilabel:`Custom Images` and select :guilabel:`Import Image`. Select :guilabel:`Import from an Object Storage URL`, then paste the `available releases <#available-releases>`_ location link with your concatenated image path into the :guilabel:`Object Storage URL` field. The URL format pasted should conform to the following:
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,273 @@
+
+    .. group-tab:: Using CLI
+    
+        The following command will directly import your image from a provided URI. You'll have the provide the values below with the exception of ``operating-system`` and ``source-image-type`` which are already provided.
```
```suggestion
        The following command will directly import your image from a provided URI. You'll have to provide the values below with the exception of ``operating-system`` and ``source-image-type`` which are already provided.
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,273 @@
+Create OKE nodes with Ubuntu Images
+-------------------------------------
+
+The following steps on creating nodes assumes you have an existing OKE cluster on Oracle Cloud, but it is not required to have existing nodes. If you don't have an OKE cluster prepared then Oracle's documentation for `creating a cluster`_ is a good place to start.
```
```suggestion
The following steps on creating nodes assume that you have an existing OKE cluster on Oracle Cloud, but it is not required to have existing nodes. If you don't have an OKE cluster prepared then Oracle's documentation for `creating a cluster`_ is a good place to start.
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,273 @@
+          --node-metadata='{"user_data": "'"$(base64 user-data.yaml)"'"}'
+
+
+View the node pool status in Oracle Cloud by navigating to :guilabel:`Kubernetes Clusters (OKE)` and choose on your cluster, then select :guilabel:`Resources` > :guilabel:`Node pools` and select the latest node pool.
```
```suggestion
View the node pool status in Oracle Cloud by navigating to :guilabel:`Kubernetes Clusters (OKE)` and choosing your cluster, then select :guilabel:`Resources` > :guilabel:`Node pools` and select the latest node pool.
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,273 @@
+
+The following instructions assume that you have configured your OKE cluster to work with self-managed nodes. If you have not done this, refer to the Oracle documentation for `working with self-managed nodes`_
+
+Before adding a self-managed node, ensure you have configured ``kubectl`` for your OKE cluster with the following command. This process will be easier if ``kubectl`` is configured for a single OKE cluster.
```
```suggestion
Before adding a self-managed node, ensure that you have configured ``kubectl`` for your OKE cluster with the following command. This process will be easier if ``kubectl`` is configured for a single OKE cluster.
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,273 @@
+
+   .. group-tab:: Using console
+  
+    Now, create the self-managed node in Oracle Cloud by navigating to :guilabel:`Compute` > :guilabel:`Instance` and select :guilabel:`Create Instance`. Next, select :guilabel:`Change Image` and select :guilabel:`My Images`, then select the Ubuntu image you recently registered. 
```
```suggestion
    Now, create the self-managed node in Oracle Cloud by navigating to :guilabel:`Compute` > :guilabel:`Instance` and select :guilabel:`Create Instance`. Next, select :guilabel:`Change Image` > :guilabel:`My Images`, and then select the Ubuntu image you recently registered. 
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,273 @@
+  
+    Now, create the self-managed node in Oracle Cloud by navigating to :guilabel:`Compute` > :guilabel:`Instance` and select :guilabel:`Create Instance`. Next, select :guilabel:`Change Image` and select :guilabel:`My Images`, then select the Ubuntu image you recently registered. 
+    
+    Setup the cloud-init for the instance by selecting :guilabel:`Show advanced options`, then select :guilabel:`Paste cloud-init script`, and then paste your completed cloud-init script (the one saved in ``user-data.yaml``).
```
```suggestion
    Setup the cloud-init for the instance by selecting :guilabel:`Show advanced options` > :guilabel:`Paste cloud-init script`, and then paste your completed cloud-init script (the one saved in ``user-data.yaml``).
```

k-dimple commented on oracle/oracle-how-to/deploy-oke-nodes-using-ubuntu-images.rst:
```diff
@@ -0,0 +1,273 @@
+Further references
+------------------
+
+For more information about oci CLI and managing self-managed nodes on your cluster, refer to the Oracle Documentation:
```
```suggestion
For more information about ``oci`` CLI and managing self-managed nodes on your cluster, refer to the Oracle Documentation:
```

k-dimple reviewed (approved):
Some minor nits remain, otherwise we are good to go, nicely done! :)
